gossiper: mutate_live_and_unreachable_endpoints: make exception safe

bhalevy opened the pull request:
Change the mutate_live_and_unreachable_endpoints procedure so that the called `func` would return a struct
`live_and_unreachable_endpoints` that contains the modified values of live and unreachable endpoints.

Those are replicated to temporary copies on all shards using `foreign<unique_ptr<>>` so that the would be automatically freed on exception.

Only after all copies are made, they are applied
on all gossiper shards in a noexcept loop
and finally, a `on_success` function is called
to apply further side effects if everything else
was replicated successfully.

The latter is still susceptible to exceptions,
but we can live with those as long as `_live_endpoints` and `_unreachable_endpoints` are synchronized on all shards.

With that, the read-only methods:
`get_live_members_synchronized` and
`get_unreachable_members_synchronized`
become trivial and they just return the required data from shard 0.

Fixes #15089

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5484/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5484/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6907/)
#### Failed Tests (1/23701):
- [test_concurrent_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5484/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_concurrent_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_restart)
#### Build Details:
- Duration: 1 hr 49 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
CI failure looks unrelated and doesn't reproduce locally for me.
Rerunning CI.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5493/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6915/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5493/)
#### Build Details:
- Duration: 1 hr 32 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
> CI failure looks unrelated and doesn't reproduce locally for me. Rerunning CI.

Please file an issue or append to one.

avikivity commented on gms/gossiper.cc:
```diff
@@ -702,10 +702,21 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
     co_await mutate_live_and_unreachable_endpoints([endpoint, &was_alive] (gossiper& g) {
-        was_alive = g._live_endpoints.erase(endpoint);
-        g._unreachable_endpoints.erase(endpoint);
+        return live_and_unreachable_endpoints{
```
Maybe mutate_live_and_unreachable_endpoints() should perform the copy and call its lambda with the copy?

Losing optimization of folding the filtering into the copy is worth the extra clarity IMO.

bhalevy replied:
That will definitely simplify the flow, and in this case which is not on the hot path maybe the optimization is indeed not worth it.

bhalevy replied:
done in v2

avikivity commented on gms/gossiper.cc:
```diff
@@ -1206,8 +1227,14 @@ version_type gossiper::get_max_endpoint_state_version(const endpoint_state& stat
+            boost::copy_range<std::unordered_map<inet_address, clk::time_point>>(g._unreachable_endpoints | boost::adaptors::filtered([endpoint] (const auto& node) {
+                return node.first != endpoint;
+            }))
+        };
```
Another case where a copy provided by mutate() would improve readability.

bhalevy replied:
done in v2

avikivity commented on gms/gossiper.hh:
```diff
@@ -248,22 +248,26 @@ private:
+    // and then applied atomcically on all shards.
+    //
+    // The `on_success` callback is called on shard 0 iff replication was successful;
+    future<> mutate_live_and_unreachable_endpoints(std::function<live_and_unreachable_endpoints(gossiper&)> func,
```
So: accept a l_a_u_e and return one (likely the same object, mutated),

bhalevy replied:
accepting a mutable reference and mutate it in place would be even simpler.

bhalevy replied:
done in v2

avikivity commented on gms/gossiper.hh:
```diff
@@ -248,22 +248,26 @@ private:
     // replicate shard 0 live and unreachable endpoints sets across all other shards.
     // _endpoint_update_semaphore must be held for the whole duration
-    future<> replicate_live_endpoints_on_change();
+    future<> replicate_live_endpoints_on_change(foreign_ptr<std::unique_ptr<live_and_unreachable_endpoints>>, uint64_t new_live_endpoints_version);
```
This pattern is so common we should figure out a good way to do it in sharded<> (perhaps with support for only creating one copy per NUMA node), but of course not in the scope of this patch.

bhalevy replied:
Agreed.  I contemplated on doing this in a generic way with shared_token_metadata and in the gossiper there's also the endpoint state map.
We should seriously consider doing that.

bhalevy commented:
> > CI failure looks unrelated and doesn't reproduce locally for me. Rerunning CI.
> 
> Please file an issue or append to one.

https://github.com/scylladb/scylladb/issues/15192#issuecomment-1874555999

bhalevy commented:
In v2 ([7b37cf2](https://github.com/scylladb/scylladb/commit/7b37cf2a10d8e0e66f62bfc2bc770803c8067a21)):
- simplified mutate_l_a_u_e by cloning `live_and_unreachable_endpoints` in outer layer and calling the lambda with a mutable ref to the cloned copy (and dropped the `gossiper&` param that isn't necessary anymore).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5517/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5517/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6931/)
#### Build Details:
- Duration: 1 hr 48 min
- Builder: spider3.cloudius-systems.com

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -853,34 +853,31 @@ future<semaphore_units<>> gossiper::lock_endpoint_update_semaphore() {
-        co_await g.replicate_live_endpoints_on_change();
-        co_return live_members;
+    return container().invoke_on(0, [] (gms::gossiper& g) {
+        return g.get_live_members();
```
This might increase test flakiness that was previously solved by introducing the synchronized() functions.

bhalevy replied:
But even if this function grabs the semaphore and returns some snapshot of live members it may change right after it releases the semaphore, so why does it matter if it loses the race with `replicate_live_endpoints_on_change`?

With the exception safety in place any value we sample is bound to have been replicated to all shards at one point in time.

kbr-scylla replied:
Yes it might change immediately, but the returned value has been observed by other shards **before** that, which is the whole point.

IIRC the original scenario that motivated the `synchronized()` functions was something as follows:
a test was restarting some node X and then starting removenode operation on another node before all nodes noticed that X is up. In the middle of streaming nodes would notice that X is UP, triggering gossiper notifications, which interrupted streaming (`stream_manager::on_restart` closes existing streaming sessions). So the test was modified to wait for nodes to see that X is up before triggering removenode. However, that didn't help, because the live nodes endpoint returned data from one shard only (shard 0), and the problem still occured, because other shards would notice that the endpoint is dead after removenode was triggered and close streaming sessions... The solution was to ensure that all shards see the new alive endpoints before triggering removenode.

kbr-scylla replied:
> With the exception safety in place any value we sample is bound to have been replicated to all shards at one point in time.

Not have been, it might be still happening.

bhalevy replied:
The reason we needed the "synchronized" functions was that the state of live and unreachable endpoint was stale on non-zero shards AFAIK.  I think that if it can provide monotonic progression of the state it should be good enough for the api users.

bhalevy replied:
That said, I don't mind add an acquisition of the semaphore for extra safety.

kbr-scylla replied:
> The reason we needed the "synchronized" functions was that the state of live and unreachable endpoint was stale on non-zero shards AFAIK

Yes and this PR brings back that problem. The returned value might not have been replicated to non-zero shards yet i.e. state on non-zero shards might be stale.

bhalevy replied:
Followup PR: https://github.com/scylladb/scylladb/pull/16629
